src/main/scala/firrtl/transforms/InferResets.scala

```diff
@@ -156,54 +177,53 @@ class InferResets extends Transform {
             val altMap = new DriverMap
             onStmt(conMap)(con)
             onStmt(altMap)(alt)
-            // Default to outerscope if not found on either side
-            val conLookup = conMap.orElse(map).lift
-            val altLookup = altMap.orElse(map).lift
             for (key <- conMap.keys ++ altMap.keys) {
-              val values = conLookup(key).getOrElse(Nil) ++ altLookup(key).getOrElse(Nil)
-              map(key) = values
+              val ds = map.getOrElseUpdate(key, mutable.ListBuffer())
+              conMap.get(key).foreach(ds ++= _)
+              altMap.get(key).foreach(ds ++= _)
             }
           case other => other.foreach(onStmt(map))
         }
       }
-      val types = new DriverMap
       mod.foreach(onStmt(types))
       types
     }
-    c.modules.foldLeft(Map[ReferenceTarget, List[ResetDriver]]()) {
-      case (map, mod) => map ++ onMod(mod).mapValues(_.toList)
-    }
+    val res = new DriverMap
+    c.modules.foreach(m => onMod(res)(m))
+    res.mapValues(_.toList).toMap
   }
 
   // Determine the type driving a given ResetType
   private def resolve(map: Map[ReferenceTarget, List[ResetDriver]]): Try[Map[ReferenceTarget, Type]] = {
-    val res = mutable.Map[ReferenceTarget, Type]()
-    val errors = new Errors
-    def rec(target: ReferenceTarget): Type = {
-      res.getOrElseUpdate(target, {
-        val drivers = map.getOrElse(target, Nil)
-        val tpes = drivers.flatMap {
-          case TargetDriver(t) => Some(TypeDriver(rec(t), () => t))
-          case td: TypeDriver  => Some(td)
-          case InvalidDriver   => None
-        }.groupBy(_.tpe)
-        tpes.keys.size match {
-          // This can occur if something of type Reset has no driver
-          case 0 => InvalidDriver.defaultType
-          case 1 => tpes.keys.head
-          case _ =>
-            // Multiple types of driver!
-            errors.append(DifferingDriverTypesException(target, tpes.toSeq))
-            tpes.keys.head
-        }
-      })
-    }
-    for ((target, _) <- map) {
-      rec(target)
+    val graph = new MutableDiGraph[Node]
+    val asyncNode = Typ(AsyncResetType)
+    val syncNode  = Typ(Utils.BoolType)
+    for ((target, drivers) <- map) {
+      val v = Var(target)
+      drivers.foreach {
+        case TargetDriver(t) =>
+          val u = Var(t)
+          graph.addPairWithEdge(v, u)
+          graph.addPairWithEdge(u, v)
+        case TypeDriver(tpe, _) =>
+          // Use nodes we already made, saves memory
+          val u = tpe match {
+            case AsyncResetType => asyncNode
+            case Utils.BoolType => syncNode
+            case other          => throwInternalError(s"Shouldn't have $other here")
+          }
+          graph.addPairWithEdge(u, v)
```

Comment:
This "fake" edge to the type seems like a clever way to share the work for the inference with the check _without_ interspersing check code into the connection edge creation.